xml/resources-3.1.rng

```diff
@@ -0,0 +1,401 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<grammar xmlns="http://relaxng.org/ns/structure/1.0"
+         datatypeLibrary="http://www.w3.org/2001/XMLSchema-datatypes">
+  <start>
+    <ref name="element-resources"/>
+  </start>
+
+  <define name="element-resources">
+    <element name="resources">
+      <zeroOrMore>
+        <choice>
+          <ref name="element-primitive"/>
+          <ref name="element-template"/>
+          <ref name="element-group"/>
+          <ref name="element-clone"/>
+          <ref name="element-master"/>
+          <ref name="element-bundle"/>
+        </choice>
+      </zeroOrMore>
+    </element>
+  </define>
+
+  <!--
+   see upgrade-2.10.xsl
+   - cibtr:table for="resource-meta-attributes"
+   -->
+  <define name="primitive-template.meta_attributes.nvpair.name-unsupported">
```

Comment:
It actually is a good idea to drop the meta-attributes limitations as well. Users can use arbitrary meta-attributes for their own administrative data that is not supported by the agent. (This surprised me, too, the first time I saw it in the wild, but it is actually useful. One day we may move away from using meta-attributes for pacemaker's purposes at all, rather using structured XML tagging for that purpose, but we would still support meta-attributes for this reason.)

Reply:
Wait, so anyone dares to repeat the inverse of the same mistake pacemaker
previously and unwisely (if it wasn't some kind of legacy thing) did with
combining it's own hints with regular user-facing configuration of the resource,
i.e. something that only v2.0 finally reconciles?

Polluting pacemaker's own annotation layer over resources is _very bad idea_,
and if it's the problem with the unchangeable resource agent, luckily the
workaround (maybe too strong expression, I still don't grasp the use case)
is as trivial as having a shell redirecting `meta-data` invocation to the
customized version...

Reply:
> Wait, so anyone dares to repeat the inverse of the same mistake pacemaker
> previously and unwisely (if it wasn't some kind of legacy thing) did with
> combining it's own hints with regular user-facing configuration of the resource,
> i.e. something that only v2.0 finally reconciles?

Yes -- the issue is that there are three namespaces but only two elements in the schema to express them. The three namespaces are: pacemaker's control parameters; the resource's control parameters; and the user's administrative data.

> Polluting pacemaker's own annotation layer over resources is a very bad idea,

Sure, there are many bad ideas in Pacemaker. :-)

The point is it's currently supported, and we don't want to break that. A future redesign would be great but isn't happening now.

> and if it's the problem with the unchangeable resource agent, luckily the
> workaround (maybe too strong expression, I still don't grasp the use case)
> is as trivial as having a shell wrapper redirecting meta-data invocation to
> the customized version...

No, it's not an extension of the resource agent parameters, it's a separate domain altogether.

As one example, let's say a user's IT department hosts a pacemaker cluster with a variety of virtual machines (including filesystem, VirtualDomain, and possibly other resources) for other departments in the company. The IT department uses a custom meta-attribute to identify which department each resource is for. This then can be used both in the cluster (via rules with value-source=meta) and outside the cluster (custom accounting software, etc.).

It's a useful feature, only the implementation is wonky. The long-term goal is to reserve meta_attributes for exactly this purpose, and have pacemaker use something structured, e.g. a new `<resource_control>` block containing e.g. `<remote node=NAME address=IP port=PORT>` and so on. But that is a massive change that will have to wait for Pacemaker 3.0 or whatever.

Reply:
As an aside, the structured approach has the obvious advantage that incorrect usage can be weeded out at schema validation time, including value typing, and it's potentially more readable. The only downside is that the schema would require changes more frequently.

Reply:
> The point is it's currently supported, and we don't want to break that.
> A future redesign would be great but isn't happening now.

Was this use case ever formally approved?
If not, it's valid to break it, it was simply a misbelieve it's OK to do that.

Reply:
> Was this use case ever formally approved?
> If not, it's valid to break it, it was simply a misbelieve it's OK to do that.

Yes, it's likely undocumented, but it's considered a valid use. And again, it's a very useful feature that can't be met otherwise -- it's just how we chose to implement it that's subpar, and can be corrected in the future.

Reply:
Sure, the reserved prefix approach is one possible redesign. It makes more sense in meta_attributes than instance_attributes though, b/c it's not specific to the agent. The same custom attribute is likely to be specified for resources of varying types -- it's truly "meta".

The existing approach, with no name reservations, has been in place for a decade. Regardless of its disadvantages, it will be supported for the near future.

Reply:
> It makes more sense in meta_attributes than instance_attributes though, b/c
> it's not specific to the agent.

I've slept on this, read the original motivations at
https://github.com/ClusterLabs/pacemaker/pull/1208
and now I agree with the affinity, but still am opposed to any kind of loose
intermixing, hence

https://github.com/ClusterLabs/pacemaker/pull/1523

The plan would not be to forbid that completely with 3.* schemas,
however, but to:
- extend `upgrade-2.10.xsl` (`entry`?) with a warning pointing out that
  custom meta-attributes should really follow the naming convention
  set forth in #1523
- possibly have C code XML sanity checking emit the same
  (or solely instead of the transformation?)

Then, I am OK to drop the restrictions from `resources-3.1.rng`
(as that's handled with the said check+warning).

Reply:
Whether the restrictions stay or go isn't that important, but it's kind of silly to have a 3.1 schema just to drop one and not the other. We could just leave it at 3.0